source/core/index.ts

```diff
@@ -0,0 +1,1472 @@
+export type BeforeErrorHook = (error: RequestError) => Promisable<RequestError>;
+
+export interface Hooks {
+	init?: InitHook[];
+	beforeRequest?: BeforeRequestHook[];
+	beforeRedirect?: BeforeRedirectHook[];
+	beforeError?: BeforeErrorHook[];
+}
+
+export type HookEvent = 'init' | 'beforeRequest' | 'beforeRedirect' | 'beforeError';
+
+export const knownHookEvents: HookEvent[] = ['init', 'beforeRequest', 'beforeRedirect', 'beforeError'];
+
+type AcceptableResponse = IncomingMessage | ResponseLike;
+type AcceptableRequestResult = AcceptableResponse | ClientRequest | Promise<AcceptableResponse | ClientRequest> | undefined;
+
+export type RequestFunction = (url: URL, options: RequestOptions, callback?: (response: AcceptableResponse) => void) => AcceptableRequestResult;
+
+export type Headers = Record<string, string | string[] | undefined>;
+
+export interface Options extends URLOptions, SecureContextOptions {
+	request?: RequestFunction;
+	agent?: Agents | false;
+	decompress?: boolean;
+	timeout?: Delays | number;
+	prefixUrl?: string | URL;
+	body?: unknown;
+	form?: {[key: string]: any};
+	json?: {[key: string]: any};
+	url?: string | URL;
+	cookieJar?: PromiseCookieJar | ToughCookieJar;
+	ignoreInvalidCookies?: boolean;
+	encoding?: BufferEncoding;
+	searchParams?: string | {[key: string]: string | number | boolean | null} | URLSearchParams;
+	dnsCache?: CacheableLookup | boolean;
+	context?: object;
+	hooks?: Hooks;
+	followRedirect?: boolean;
+	maxRedirects?: number;
+	cache?: string | CacheableRequest.StorageAdapter | false;
+	throwHttpErrors?: boolean;
+	username?: string;
+	password?: string;
+	http2?: boolean;
+	allowGetBody?: boolean;
+	lookup?: CacheableLookup['lookup'];
+	rejectUnauthorized?: boolean;
+	headers?: Headers;
+	methodRewriting?: boolean;
+
+	// From http.RequestOptions
+	localAddress?: string;
+	socketPath?: string;
+	method?: string;
+	createConnection?: (options: http.RequestOptions, oncreate: (error: Error, socket: Socket) => void) => Socket;
+}
+
+export interface NormalizedOptions extends Options {
+	method: Method;
+	url: URL;
+	timeout: Delays;
+	prefixUrl: string;
+	ignoreInvalidCookies: boolean;
+	decompress: boolean;
+	searchParams?: URLSearchParams;
+	cookieJar?: PromiseCookieJar;
+	headers: Headers;
+	context: object;
+	hooks: Required<Hooks>;
+	followRedirect: boolean;
+	maxRedirects: number;
+	cache?: string | CacheableRequest.StorageAdapter;
+	throwHttpErrors: boolean;
+	dnsCache?: CacheableLookup;
+	cacheableRequest?: (options: string | URL | http.RequestOptions, callback?: (response: http.ServerResponse | ResponseLike) => void) => CacheableRequest.Emitter;
+	http2: boolean;
+	allowGetBody: boolean;
+	rejectUnauthorized: boolean;
+	lookup?: CacheableLookup['lookup'];
+	methodRewriting: boolean;
+	username: string;
+	password: string;
+	[kRequest]: HttpRequestFunction;
+	[kIsNormalizedAlready]?: boolean;
+}
+
+export interface Defaults {
```

Comment:
We shouldn't extend `NormalizedOptions` nor `Options` to require strict (aka "full", "complete") values.